File: LogisimPlus/LogisimPlus/logisim/prefs/PrefMonitorStringOpts.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.prefs
{

	internal class PrefMonitorStringOpts : AbstractPrefMonitor<string>
	{
		private string[] opts;
		private string value;
		private string dflt;

		internal PrefMonitorStringOpts(string name, string[] opts, string dflt) : base(name)
		{
			this.opts = opts;
			this.value = opts[0];
			this.dflt = dflt;
			Preferences prefs = AppPreferences.Prefs;
			set(prefs.get(name, dflt));
			prefs.addPreferenceChangeListener(this);
		}

		public override string get()
		{
			return value;
		}

		public virtual void set(string newValue)
		{
			string oldValue = value;
			if (!isSame(oldValue, newValue))
			{
				AppPreferences.Prefs.put(Identifier, newValue);
			}
		}

		public override void preferenceChange(PreferenceChangeEvent @event)
		{
			Preferences prefs = @event.getNode();
			string prop = @event.getKey();
			string name = Identifier;
			if (prop.Equals(name))
			{
				string oldValue = value;
				string newValue = prefs.get(name, dflt);
				if (!isSame(oldValue, newValue))
				{
					string[] o = opts;
					string chosen = null;
					for (int i = 0; i < o.Length; i++)
					{
						if (isSame(o[i], newValue))
						{
							chosen = o[i];
							break;
						}
					}
					if (string.ReferenceEquals(chosen, null))
					{
						chosen = dflt;
					}
					value = chosen;
					AppPreferences.firePropertyChange(name, oldValue, chosen);
				}
			}
		}

		private static bool isSame(string a, string b)
		{
			return string.ReferenceEquals(a, null) ? string.ReferenceEquals(b, null) : a.Equals(b);
		}
	}

}

File: logisim/app/src/main/java/logisim/prefs/PrefMonitorStringOpts.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.prefs;

import java.util.Objects;
import java.util.prefs.PreferenceChangeEvent;
import java.util.prefs.Preferences;

class PrefMonitorStringOpts extends AbstractPrefMonitor<String> {
	private String[] opts;
	private String value;
	private String dflt;

	PrefMonitorStringOpts(String name, String[] opts, String dflt) {
		super(name);
		this.opts = opts;
		value = opts[0];
		this.dflt = dflt;
		Preferences prefs = AppPreferences.getPrefs();
		set(prefs.get(name, dflt));
		prefs.addPreferenceChangeListener(this);
	}

	public String get() {
		return value;
	}

	public void set(String newValue) {
		String oldValue = value;
		if (!isSame(oldValue, newValue)) AppPreferences.getPrefs().put(getIdentifier(), newValue);
	}

	public void preferenceChange(PreferenceChangeEvent event) {
		Preferences prefs = event.getNode();
		String prop = event.getKey();
		String name = getIdentifier();
		if (prop.equals(name)) {
			String oldValue = value;
			String newValue = prefs.get(name, dflt);
			if (!isSame(oldValue, newValue)) {
				String[] o = opts;
				String chosen = null;
				for (String s : o)
					if (isSame(s, newValue)) {
						chosen = s;
						break;
					}
				if (chosen == null)
					chosen = dflt;
				value = chosen;
				AppPreferences.firePropertyChange(name, oldValue, chosen);
			}
		}
	}

	private static boolean isSame(String a, String b) {
		return Objects.equals(a, b);
	}
}
